[TensorExpr] Add methods for inspecting generated code in `TensorExprKernel`.

ZolotukhinM opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#57074 [TensorExpr] Add methods for inspecting generated code in `TensorExprKernel`.**
* #57382 [TensorExpr] Remove `dtype_` and add `buf_` fields to `CodeGen::BufferArg`.
* #57169 [JIT] Add a pass for removing a first (self) argument from a graph if it is unused.
* #57076 [JIT] Add a pass for annotating graph with input types derived from sample inputs.

The new methods allow to peak into bufferArgs which describe parameters
that codegen expects. This description includes info whether a given
parameter is a scalar var or a buffer and in case it's a buffer allows
to get the corresponding `Buf*` pointer from which we could get the
expected sizes.

Differential Revision: [D28048289](https://our.internmc.facebook.com/intern/diff/D28048289)

Commit:
add getConstantDescriptors, add a test trying to retrieve generated asm code

[ghstack-poisoned]

Commit:
Update on "add getConstantDescriptors, add a test trying to retrieve generated asm code"




[ghstack-poisoned]

Commit:
Update on "add getConstantDescriptors, add a test trying to retrieve generated asm code"




[ghstack-poisoned]

Commit:
Update on "add getConstantDescriptors, add a test trying to retrieve generated asm code"


Differential Revision: [D28048289](https://our.internmc.facebook.com/intern/diff/D28048289)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add methods for inspecting generated code in `TensorExprKernel`."


The new methods allow to peak into bufferArgs which describe parameters
that codegen expects. This description includes info whether a given
parameter is a scalar var or a buffer and in case it's a buffer allows
to get the corresponding `Buf*` pointer from which we could get the
expected sizes.

Differential Revision: [D28048289](https://our.internmc.facebook.com/intern/diff/D28048289)

[ghstack-poisoned]

bertmaher commented on torch/csrc/jit/tensorexpr/kernel.h:
```diff
@@ -75,6 +75,14 @@ class TORCH_API TensorExprKernel {
     return graph_;
   }
 
+  std::vector<ConstantDescr> getConstantDescriptors() const {
+    return constants_;
+  }
+
+  std::vector<CodeGen::BufferArg> getBufferArgs() const {
+    return bufferArgs_;
```
Should these return `const std::vector<>&`?

ZolotukhinM replied:
probably yes :) good catch!

bertmaher reviewed (approved):
One small comment, otherwise lgtm

Commit:
Update on "[TensorExpr] Add methods for inspecting generated code in `TensorExprKernel`."


The new methods allow to peak into bufferArgs which describe parameters
that codegen expects. This description includes info whether a given
parameter is a scalar var or a buffer and in case it's a buffer allows
to get the corresponding `Buf*` pointer from which we could get the
expected sizes.

Differential Revision: [D28048289](https://our.internmc.facebook.com/intern/diff/D28048289)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add methods for inspecting generated code in `TensorExprKernel`."


The new methods allow to peak into bufferArgs which describe parameters
that codegen expects. This description includes info whether a given
parameter is a scalar var or a buffer and in case it's a buffer allows
to get the corresponding `Buf*` pointer from which we could get the
expected sizes.

Differential Revision: [D28048289](https://our.internmc.facebook.com/intern/diff/D28048289)

[ghstack-poisoned]

mruberry commented:
Unlanding; this appears to be breaking pytorch_linux_xenial_py3_clang5_asan_test2:

```
May 04 15:38:10 [ RUN      ] Kernel.CodegenInspection
May 04 15:38:10 unknown file: Failure
May 04 15:38:10 C++ exception with description "Expected to find ".text" but did not find it
May 04 15:38:10 Searched string:
May 04 15:38:10 From CHECK: .text
May 04 15:38:10 " thrown in the test body.
May 04 15:38:10 [  FAILED  ] Kernel.CodegenInspection (4 ms)
```

ljk53 commented on torch/csrc/jit/tensorexpr/kernel.h:
```diff
@@ -85,6 +85,14 @@ class TORCH_API TensorExprKernel {
     return graph_;
   }
 
+  const std::vector<ConstantDescr>& getConstantDescriptors() const {
```
@ZolotukhinM in the current prototype we serialize the constant parameters as IValue wrapped tensors in the compiled model file: https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/mobile/nnc/context.h#L131

Looks like ConstantDescr contains Buf which has dims (represented as Expr - they are constant, right?) and dtype - so it's possible to reconstruct the tensors and serialize, right? Or shall we directly wrap the ConstantDescr and the raw bytes into IValue?

Another option is to expose a way to inspect the link to the original IR node. It might allow us to persist things represented in the form of opaque tensors used by external ops (for which NNC generated assembly could simply pass through).

ZolotukhinM replied:
> ConstantDescr contains Buf which has dims (represented as Expr - they are constant, right?)

This is a more general case and Exprs do not have to always be constant. However, in our case we can add an assert that they are and then convert to `vector<int>` to construct `InputDesc` or something similar if we need it. When the shapes are static we expect this exprs to be constant. When we extend to support dynamic shapes that assert would need to be relaxed, but the interface to query the info about tensors would stay the same and we will already have a way to represent symbolic shapes (with Var expressions).
